crates/polars-mem-engine/src/executors/scan/ipc.rs

```diff
@@ -65,211 +65,109 @@ impl IpcExec {
 
         let projection = materialize_projection(
             self.file_options.with_columns.as_deref(),
-            &self.schema,
+            &self.file_info.schema,
             None,
             self.file_options.row_index.is_some(),
         );
 
-        let n_rows = self
-            .file_options
-            .n_rows
-            .map(|n| IdxSize::try_from(n).unwrap());
-
-        let row_limit = n_rows.unwrap_or(IdxSize::MAX);
+        let read_path = |path_index: usize, n_rows: Option<usize>| {
+            IpcReader::new(path_idx_to_file(path_index)?)
+                .with_n_rows(n_rows)
+                .with_row_index(self.file_options.row_index.clone())
+                .with_projection(projection.clone())
+                .with_hive_partition_columns(
+                    self.hive_parts
+                        .as_ref()
+                        .map(|x| x[path_index].materialize_partition_columns()),
+                )
+                .memory_mapped(
+                    self.options
+                        .memory_map
+                        .then(|| self.paths[path_index].clone()),
+                )
+                .finish()
+        };
 
-        // Used to determine the next file to open. This guarantees the order.
-        let path_index = AtomicUsize::new(0);
-        let row_counter = RwLock::new(ConsecutiveCountState::new(self.paths.len()));
+        let mut dfs = if let Some(mut n_rows) = self.file_options.n_rows {
```

Comment:
Use sequential read if there is a row limit - this is usually what we do in other places (union, csv scan)